More generic events and remove the need of CRTP for some classes

herpec-j opened the pull request:
- Delete BaseReceiver and BaseSystem
- Delete BaseEvent and Event classes ==> We can now emit every kind of data, even int or float for example
- Remove the need of CRTP for Receiver and System classes
- Possibility to specify another method than "receive" while subscribing to an event (it will use the receive method by default if no method is provided so that it's not a breaking change)
- Constructor and Destructor of EventManager class automatically generated ==> Delete file Event.cc
- Destructor of System class automatically generated
- Fix compilation of benchmarks_test on Visual Studio 12

alecthomas commented:
This is really nice, thank you. However, there are two major issues:

1. I have removed (generic) events and systems from the [v2 API](https://github.com/alecthomas/entityx/tree/2.x).
2. It breaks backwards compatibility.

The first is not necessarily a deal breaker as I'm not 100% sure that removing systems and events is the right way forward. But the second is a bit of a problem.

Also, I don't think this:

```c++
ev.subscribe<EntityDestroyedEvent>(receiver, &EntityDestroyedEventReceiver::entityDestroyed);
```

is an improvement over this for the default case (even given the removal of CRTP):

```c++
ev.subscribe<EntityDestroyedEvent>(receiver);
```

It should be possible to overload receive to keep the old behaviour. Something like:

```
template <typename T>
void subscribe(std::function<void(const T&)> f) {
  // ...
}

template <typename T, typename R>
void subscribe(R &receiver) {
  auto method = static_cast<void(R::*)(const T&)>(&R::receive);
  subscribe<T>(std::bind(method, &receiver, std::placeholders::_1));
}

subscribe<int>(receiver);
```

I'll put some other comments inline.

alecthomas commented on entityx/Event.h:
```diff
@@ -102,18 +65,26 @@ class Receiver : public BaseReceiver {
+	// Method used to get the default callback function for an event (receive method).
+	// Directly put &Receiver::receive as the default parameter of the subscribe function doesn't work with VS12.
+   template <typename E, typename Receiver>
+   static CallbackSignature<E, Receiver> getDefaultCallbackFunction() {
```
The naming convention I use in EntityX is for methods/functions to be `underscore_separated_lowercase`. Please fix here and elsewhere (eg. in tests).

alecthomas commented on entityx/Event.h:
```diff
@@ -139,21 +110,21 @@ class EventManager : entityx::help::NonCopyable {
-    assert(base.connections_.find(Event<E>::family()) != base.connections_.end());
-    auto pair = base.connections_[Event<E>::family()];
+	assert(base.connections_.find(UIDGenerator::GetUID<E>()) != base.connections_.end());
+	auto pair = base.connections_[UIDGenerator::GetUID<E>()];
```
You appear to have introduced tabs. Please remove them and use spaces only.

alecthomas commented on entityx/help/UIDGenerator.h:
```diff
@@ -0,0 +1,34 @@
+  ~UIDGenerator(void) = delete;
+
+  template <typename T>
+  static typename std::enable_if<!std::is_const<T>::value && !std::is_volatile<T>::value, std::size_t>::type GetUID(void) {
```
Same comment as above with naming conventions: `get_uid()`, `get_counter_and_increment()`, and so on.

herpec-j commented:
Thank you for the feedback!

I fixed everything related to tabulations and naming conventions, it was definitely a bad thing.
I also modified the subscribe method in order to take a std::function as a parameter, the problem however is that if we don't receive a instance of Receiver as a parameter, we can't keep a trace of the new connection in order to unsubscribe later.
Finally, for the removal of CRTP classes, I thought it would be a good idea to remove them as their are not needed, but that indeed breaks backward compatibility, unfortunately. Maybe it wasn't such a good idea to remove them finally.

alecthomas commented:
Thank you for the PR, I appreciate it, but I don't think I can merge this with the backwards incompatibility.
